### PR TITLE
Introduce soft limit in time management

### PR DESCRIPTION
**Pull request by XInTheDark**
Time management now uses both `opt_time` (the soft time limit) and `max_time` (the hard time limit). Every 256 nodes, we check if the hard limit is reached; every 16384 nodes, we check if the soft limit is reached, as well as recalculate the optimal time using search results. A formula for calculating the hard limit was added, and the previous formula is now used to calculate the soft limit.

Currently, two heuristics are used to adjust the time usage in each iteration: falling eval and best move instability.

Lastly, also made a lot of time management params tunable.

Passed STC: https://montychess.org/tests/view/668f276142a4028e60f12274
LLR: 8.06 (-2.94,2.94) <0.00,4.00>
Total: 7744 W: 2204 L: 1749 D: 3791
Ptnml(0-2): 103, 737, 1794, 1078, 160 

Passed LTC: https://montychess.org/tests/view/668f5f7b5912be719de319e0
LLR: 2.94 (-2.94,2.94) <1.00,5.00>
Total: 7470 W: 1852 L: 1673 D: 3945
Ptnml(0-2): 64, 805, 1836, 948, 82 

Bench: 2149456

### CODE REVIEW

File: src/chess/moves.rs [hunk 1]
```diff
@@ -25,6 +25,12 @@ impl std::fmt::Display for Move {
     }
 }
 
+impl PartialEq for Move {
+    fn eq(&self, other: &Self) -> bool {
+        self.mov == other.mov
+    }
+}
+
```
Comment (jw1912) [inline src/chess/moves.rs]: You can derive this on the struct.

File: src/mcts.rs [hunk 2]
```diff
@@ -93,23 +98,61 @@ impl<'a> Searcher<'a> {
...
+            if nodes & 255 == 0 {
                 if self.abort.load(Ordering::Relaxed) {
                     break;
                 }
 
+                let elapsed = timer.elapsed().as_millis();
+
                 if let Some(time) = limits.max_time {
-                    if timer.elapsed().as_millis() >= time {
+                    if elapsed >= time {
```
Comment (jw1912) [inline src/mcts.rs]: Can you undo this change?

File: src/mcts.rs [hunk 3]
```diff
@@ -288,6 +331,16 @@ impl<'a> Searcher<'a> {
         (pv, score)
     }
 
+    fn get_best_move(&self) -> Move {
+        let idx = self.tree.get_best_child(self.tree.root_node());
+        let action = self.tree.edge(self.tree.root_node(), idx);
+        Move::from(action.mov())
```
Comment (jw1912) [inline src/mcts.rs:335]: This is unused, can you either use it or remove it?

File: src/mcts/helpers.rs [hunk 4]
```diff
@@ -60,35 +60,55 @@ impl SearchHelpers {
...
+        increment: Option<u64>,
+        ply: u32,
+        movestogo: Option<u64>,
+        params: &MctsParams,
+    ) -> (u128, u128) {
         let inc = increment.unwrap_or(0);
 
-        let mut max_time = if let Some(mtg) = movestogo {
+        let opt_time;
+        let max_time;
```
Comment (jw1912) [inline src/mcts/helpers.rs]: You don't need to do this, in the block below you can just directly return the tuple from the function:

```rust
        if let Some(mtg) = movestogo {
            // Cyclic time control (x moves in y seconds)
            let time = (time as f64 / (mtg as f64).clamp(1.0, 30.0)) as u128;
            (time, time)
        } else {
            // Increment time control (x seconds + y increment)
            let mtg = 30;

            // blah blah ...

            let opt_time = (opt_scale * bonus * time_left) as u128;
            let max_time =
                (max_scale * opt_time as f64).min(time as f64 * params.tm_max_time()) as u128;

            (opt_time, max_time)
        }
```

File: src/mcts/helpers.rs [hunk 5]
```diff
@@ -60,35 +60,55 @@ impl SearchHelpers {
...
     /// need for it here.
-    pub fn get_time(time: u64, increment: Option<u64>, ply: u32, movestogo: Option<u64>) -> u128 {
+    pub fn get_time(
+        time: u64,
+        increment: Option<u64>,
+        ply: u32,
+        movestogo: Option<u64>,
+        params: &MctsParams,
+    ) -> (u128, u128) {
         let inc = increment.unwrap_or(0);
```
Comment (jw1912) [inline src/mcts/helpers.rs]: This can be moved into the `else` block

File: src/mcts.rs [hunk 6]
```diff
@@ -288,6 +331,16 @@ impl<'a> Searcher<'a> {
...
+    }
+
+    fn get_cp(&self, score: f32) -> f32 {
```
Comment (jw1912) [inline src/mcts.rs]: This doesn't need to be a method. Remove the `&self` and you can call it with `Searcher::get_cp(score)`.

File: src/mcts/params.rs [hunk 7]
```diff
@@ -115,4 +144,19 @@ make_mcts_params! {
...
+    tm_optscale_value3: f64 = 0.50, 0.1, 1.0, 0.05, 0.002;
+    tm_optscale_value4: f64 = 0.25, 0.1, 1.0, 0.02, 0.002;
+    tm_max_value1: f64 = 3.39, 1.0, 10.0, 0.4, 0.002;
+    tm_max_value2: f64 = 3.01, 1.0, 10.0, 0.4, 0.002;
+    tm_max_value3: f64 = 2.93, 1.0, 10.0, 0.4, 0.002;
+    tm_maxscale_value1: f64 = 12.0, 1.0, 24.0, 1.2, 0.002;
+    tm_maxscale_value2: f64 = 4.00, 1.0, 12.0, 0.6, 0.002;
+    tm_bonus_ply: f64 = 11.0, 1.0, 30.0, 1.5, 0.002;
+    tm_bonus_value1: f64 = 0.5, 0.1, 2.0, 0.05, 0.002;
+    tm_max_time: f64 = 0.850, 0.400, 0.990, 0.085, 0.002;
```
Comment (jw1912) [inline src/mcts/params.rs:161]: I'm not sure about having these TM params always exposed as tunable, am open to being convinced otherwise.

File: src/mcts.rs [hunk 8]
```diff
@@ -93,23 +98,61 @@ impl<'a> Searcher<'a> {
 
             nodes += 1;
 
-            if nodes % 128 == 0 {
+            if nodes & 255 == 0 {
                 if self.abort.load(Ordering::Relaxed) {
```
Comment (jw1912) [inline src/mcts.rs:102]: Same goes for the other instances of this, please just do `nodes % 256`, it conveys the intent clearer and the optimisation is done even in debug mode.

**Review by jw1912 — CHANGES_REQUESTED** [review]

File: src/mcts/params.rs (hunk 7 above)
Comment (XInTheDark) [reply src/mcts/params.rs:161]: Since I'm planning to do a timeman tune very soon, I think the params can be removed after the tune instead. currently the params aren't tuned at all and so there should almost certainly be room for gains.

File: src/mcts.rs (hunk 3 above)
Comment (XInTheDark) [reply src/mcts.rs:335]: It's used to update the best_move_changes variable: 
```rust
let new_best_move = self.get_best_move();
if new_best_move != best_move {
    best_move = new_best_move;
    best_move_changes += 1;
}
```

Comment (jw1912) [reply src/mcts.rs:335]: @XInTheDark ah, I see, in that case can you also use it at the end of the search function?

File: src/mcts/params.rs (hunk 7 above)
Comment (jw1912) [reply src/mcts/params.rs:161]: @XInTheDark cool, I'm happy with that.

File: src/mcts.rs (hunk 3 above)
Comment (XInTheDark) [reply src/mcts.rs:335]: > @XInTheDark ah, I see, in that case can you also use it at the end of the search function?

Apparently the search function needs to return both best move and eval, and get_best_move() only returns best move as a speedup. Although I think we can just do get_pv(0) instead?

Comment (jw1912) [reply src/mcts.rs:335]: Nah no worries then

**Review by jw1912 — APPROVED** [review]
Very nice
